src/main/src/main/java/org/geoserver/catalog/ResourcePool.java

```diff
@@ -1111,7 +1111,13 @@ private FeatureType buildFeatureType(
                 }
             }
             ft = tb.buildFeatureType();
+            // extension point for retyping the feature type
+            for (RetypeFeatureTypeCallback callback :
+                    GeoServerExtensions.extensions(RetypeFeatureTypeCallback.class)) {
+                ft = callback.retypeFeatureType(info, ft);
```

Comment:
protect yourself with a try catch here so you can clearly blame any implementation that causes problems